contracts/ProcedureTable.sol

```diff
@@ -232,158 +268,428 @@ library ProcedureTable {
-        _serialiseProcedure(p, 0, pPointer);
-    }
-
-    function _serialiseProcedure(Procedure memory p, uint8 storagePage, uint248 pPointer) internal {
-        // Store the keyIndex at this location
-        _set(storagePage, pPointer, p.keyIndex);
-        // Store the address at the loction after this (making this data 2
-        // uint256 wide).
-        _set(storagePage, pPointer + 1, uint256(p.location));
-
-        // The first value of the array is the number of capabilities
-        _set(storagePage, pPointer + 2, p.caps.length);
-        _serialiseCapArray(p, storagePage, pPointer);
-    }
-
-    function _serialiseCapArray(Procedure memory p, uint8 storagePage, uint248 pPointer) internal {
-        // n is the storage key index
-        uint248 n = 0;
-        // i is the index of the cap
-        for (uint248 i = 0; i < p.caps.length; i++) {
-            uint256 nValues = p.caps[i].values.length;
-            uint256 capSize = nValues + 1;
-            _set(storagePage, pPointer + 3 + n, capSize); n++;
-            _set(storagePage, pPointer + 3 + n, p.caps[i].capType); n++;
-            for (uint248 j = 0; j < nValues; j++) {
-                _set(storagePage, pPointer + 3 + n, p.caps[i].values[j]); n++;
+        uint256 pPointer = _getProcedurePointerByKey(key);
+        _serialiseProcedure(p, pPointer, caps);
+    }
+
+    function _serialiseProcedure(Procedure memory p, uint256 pPointer, uint256[] caps) internal {
+        // Store the address of the code contract
+        _set(pPointer + 0, uint256(p.location));
+        // Store the keyIndex
+        _set(pPointer + 1, p.keyIndex);
+        _serialiseCapArray(pPointer, caps);
+    }
+
+    function _serialiseCapArray(uint256 pPointer, uint256[] caps) internal {
+        uint192 currentProcedure = _getCurrentProcedure();
+        // If their is no current procedure, we can do anything.
+        // TODO: this is something for consideration, we often ask the kernel to
+        // register things directly.
+
+        // Variables for later use
+        uint256 currentLength;
+        uint256 val;
+        uint256 j;
+        uint256 thisTypeLength;
+        // i is the index of the key value
+        for (uint256 i = 0; (i+2) < caps.length; ) {
+            uint256 capSize = caps[i+0];
+            uint256 capType = caps[i+1];
+            // TODO: check for overflows
+            uint8 capIndex = uint8(caps[i+2]);
+
+            if (capSize == 3) {
+                // If the capSize is 3, we don't need to create a subset, we just
+                // copy whatever is at capIndex
+
+                // We can't deal with this properly yet, but for the pursposes of setEntry, we will just set an empty value
+                currentLength = _get(pPointer | (capType*0x10000));
+                // Increment length
+                _set(pPointer | (capType*0x10000), currentLength + 1);
+
+                if (currentProcedure == 0) {
```

Comment:
This means we need to make sure that `0` is not a valid `procedureId` value. Is this as part of bootstrapping?

Reply:
This is not a requirement from me, so unless you have a good reason to keep it, less just remove the requirement. Less we have to think about.

Reply:
Let's keep it for now.